### PR TITLE
Support SINK INTO in JDBC SQL connector [HZ-1639]

### PR DESCRIPTION
**Pull request by orcunc**
Currently, when an existing entry is updated in the GenericMapStore two SQL statements are executed. 
1. Insert statement. This  throws an SQLException
2. Upon receiving the SQLException Update statement

However some databases support upsert statements, which inserts or deletes within a single operation.
Therefore the GenericMapStore needs to use "SINK INTO ..." statement and this statement needs to be converted into SQL dialect for upsert operation.

Jira : https://hazelcast.atlassian.net/browse/HZ-1639

Related : https://github.com/hazelcast/hazelcast/issues/22738

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible

### CODE REVIEW

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpsertProcessorSupplier.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        SQLException next = e.getNextException();
+        return e instanceof SQLNonTransientException
+                || e.getCause() instanceof SQLNonTransientException
+                || !isTransientCode(e.getSQLState())
+                || (next != null && e != next && isNonTransientException(next));
+    }
+
+    private boolean isTransientCode(String code) {
+        // Full list of error codes at:
+        // https://www.postgresql.org/docs/current/errcodes-appendix.html
```
Comment (viliam-durina) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpsertProcessorSupplier.java:95]: Isn't this postgres-specific? Are we using it for all databases?

**Review by viliam-durina — CHANGES_REQUESTED** [review]

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpsertProcessorSupplier.java (hunk 1 above)
Comment (orcunc) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpsertProcessorSupplier.java:95]: It was an example to a link. I have deleted the comment

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpsertProcessorSupplier.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    @SuppressWarnings("BooleanExpressionComplexity")
+    private boolean isNonTransientException(SQLException e) {
+        SQLException next = e.getNextException();
+        return e instanceof SQLNonTransientException
+                || e.getCause() instanceof SQLNonTransientException
+                || !isTransientCode(e.getSQLState())
+                || (next != null && e != next && isNonTransientException(next));
+    }
+
+    private boolean isTransientCode(String code) {
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpsertProcessorSupplier.java]: I have moved the isTransientCode() to a common class

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpsertProcessorSupplier.java (hunk 1 above)
Comment (orcunc) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpsertProcessorSupplier.java]: According to  @Frantisek it is postgre specific because postgre does not throw SQLNonTransientException but only throws org.postgresql.util.PSQLException

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/MySQLUpsertQueryBuilder.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    protected void getValuesClause(JdbcTable jdbcTable, StringBuilder stringBuilder) {
+
+        List<String> dbFieldNames = jdbcTable.dbFieldNames();
+
+        String values = dbFieldNames.stream()
+                .map(dbFieldName -> "?")
+                .collect(Collectors.joining(","));
+
+        String format = String.format("VALUES (%s) ", values);
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/MySQLUpsertQueryBuilder.java:53]: Ok

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/MySQLUpsertQueryBuilder.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                .append(") ");
+    }
+
+    protected void getValuesClause(JdbcTable jdbcTable, StringBuilder stringBuilder) {
+
+        List<String> dbFieldNames = jdbcTable.dbFieldNames();
+
+        String values = dbFieldNames.stream()
+                .map(dbFieldName -> "?")
+                .collect(Collectors.joining(","));
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/MySQLUpsertQueryBuilder.java:51]: They are not the same. 
The first one is giving "?,?,?" string . 
The suggested one is giving "field1,field2,field3" string . 
We need to use ? character instead of field name

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/H2UpsertQueryBuilder.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        StringBuilder stringBuilder = new StringBuilder();
+
+        getMergeClause(jdbcTable, stringBuilder);
+        getKeyClause(jdbcTable, stringBuilder);
+        getValuesClause(jdbcTable, stringBuilder);
+
+        query = stringBuilder.toString();
+    }
+
+    protected void getMergeClause(JdbcTable jdbcTable, StringBuilder stringBuilder) {
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/H2UpsertQueryBuilder.java:35]: Made package private for testing

File: extensions/mapstore/src/main/java/com/hazelcast/mapstore/Queries.java [hunk 6]
```diff
@@ -43,34 +43,31 @@ class Queries {
     private final Map<Integer, String> deleteAllQueries = new ConcurrentHashMap<>();
 
     Queries(String mapping, String idColumn, List<SqlColumnMetadata> columnMetadata) {
-        loadQuery = "SELECT * FROM \"" + mapping + "\" WHERE \"" + idColumn + "\" = ?";
+        loadQuery = String.format("SELECT * FROM \"%s\" WHERE \"%s\" = ?", mapping, idColumn);
```
[the remark this answers is not in the document]
Comment (frant-hartm) [reply extensions/mapstore/src/main/java/com/hazelcast/mapstore/Queries.java:46]: The values of these strings come from the XML / Yaml configuration. Do these really need to be escaped?

Calcite provides `org.apache.calcite.sql.SqlDialect#quoteIdentifier(java.lang.String)`, which escapes double quotes by doubling them. But I am not sure we should introduce a direct dependency on calcite here.

We don't care about different dialects here because this always runs against Hazelcast SQL.

File: extensions/mapstore/src/main/java/com/hazelcast/mapstore/Queries.java [hunk 7]
```diff
@@ -43,34 +43,31 @@ class Queries {
...
 
-        loadAllFactory = n -> "SELECT * FROM \"" + mapping + "\" WHERE \"" + idColumn + "\" IN ("
-                + queryParams(n)
-                + ")";
+        loadAllFactory = n -> String.format("SELECT * FROM \"%s\" WHERE \"%s\" IN (%s)", mapping, idColumn,
+                queryParams(n));
```
Comment (frant-hartm) [inline extensions/mapstore/src/main/java/com/hazelcast/mapstore/Queries.java:49]: Do you really find the `String.format` more readable than the string concatenation?

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/H2UpsertQueryBuilderTest.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+        when(jdbcTable.getExternalName()).thenReturn("table1");
+        when(jdbcTable.getPrimaryKeyList()).thenReturn(Arrays.asList("pk1", "pk2"));
+        when(jdbcTable.dbFieldNames()).thenReturn(Arrays.asList("field1", "field2"));
+
+        H2UpsertQueryBuilder builder = new H2UpsertQueryBuilder(jdbcTable);
+
+        String query = builder.query();
+        assertEquals("MERGE INTO table1 (field1,field2) KEY (pk1,pk2 ) VALUES (?, ?)", query);
+
```
Comment (frant-hartm) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/H2UpsertQueryBuilderTest.java:76]: nit: Our convention is no empty lines after/before beginning/end of the method.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/H2UpsertQueryBuilderTest.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    public void testQuery() {
+
+        when(jdbcTable.getExternalName()).thenReturn("table1");
+        when(jdbcTable.getPrimaryKeyList()).thenReturn(Arrays.asList("pk1", "pk2"));
+        when(jdbcTable.dbFieldNames()).thenReturn(Arrays.asList("field1", "field2"));
+
+        H2UpsertQueryBuilder builder = new H2UpsertQueryBuilder(jdbcTable);
+
+        String query = builder.query();
+        assertEquals("MERGE INTO table1 (field1,field2) KEY (pk1,pk2 ) VALUES (?, ?)", query);
```
Comment (frant-hartm) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/H2UpsertQueryBuilderTest.java:75]: The generated SQL doesn't have unified whitespace style:
-  sometimes you have no whitespace `(field1,field2)`
- sometimes you have weird one `(pk1,pk2 )`
- and sensible one  `(?, ?)`

**Review by frant-hartm — CHANGES_REQUESTED** [review]
@orcunc could you please split this into 2 PRs, one for the changes in hazelcast-sql module (you can keep those in this PR), 2nd for changes in generic-map-store. It will be easier/faster to review and result in a nicer history - smaller PRs are better.

Regarding escaping the identifiers, in the JdbcConnector and the builders there (in hazelcast-sql module) - let's use the dialect already available in JdbcTable and quote the identifiers / etc. using `org.apache.calcite.sql.SqlDialect#quoteIdentifier(java.lang.String)`

This way we will
- escape the quotes in the identifiers
- use the right quotes according to the dialect, e.g. MySQL will use ` and escape them correctly.

In the generic map store I am not sure if the escaping is needed, if you can specify the mapping name, id column in the config, you have a much higher access to the member than SQL console anyway. But maybe we should do it for the sake of correctness.

The queries from GenericMapStore run always against our SQL engine, so we don't need to worry about dialects, maybe we could provide some util class, which also our users could use to quote identifiers (and other things). Maybe we can start with internal class and later make it public @viliam-durina wdyt?

File: extensions/mapstore/src/main/java/com/hazelcast/mapstore/Queries.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (orcunc) [reply extensions/mapstore/src/main/java/com/hazelcast/mapstore/Queries.java:46]: 1. If we escape all identifiers, then MySQL needs to run in Ansi SQL mode.  With "sessionVariables=sql_mode=ANSI"  For example like this
docker run \
  -d \
  -it \ 
  -e MYSQL_ROOT_PASSWORD=root \
  -e MYSQL_DATABASE=db  \
  --name mysql \
  --rm \
  -p 3306:3306 \
  -v ${PWD}/schema.sql:/docker-entrypoint-initdb.d/schema.sql \
  mysql:8.0 --sql-mode="ANSI_QUOTES"

But most servers do not run with sql-mode ON. 

And for MySQL table and column escaping character is backtick not double quote. This works for normal servers even though column name has a hyphen (normally considered minus operator) 

CREATE TABLE foo (
  `time-taken` int(10) unsigned NOT NULL 
);

This does not **work** for a normal server
CREATE TABLE foo (
  "time-taken" int(10) unsigned NOT NULL 
);
So adding double quote for MySQL is not conventional. 

2. For PostgreSQL, the default behavior is converting everything to lower case.  If we use double quotes then PostgreSQL will use everything as it is. But the below example is going give error, because table name has to be used with double quotes all the time. And this is not the convention in PostgreSQL. People will get confused.
CREATE TABLE "People" (
 id SERIAL NOT NULL,
 email TEXT NOT NULL,
 PRIMARY KEY(id)
);

select * from people; <-- Error here

ERROR: relation "people" does not exist

See https://lerner.co.il/2013/11/30/quoting-postgresql/

3. Similar for MS SQL . Please see
https://learn.microsoft.com/en-us/sql/t-sql/statements/set-quoted-identifier-transact-sql?redirectedfrom=MSDN&view=sql-server-ver16

Bottom line : 
In my opinion we should get rid of all quotes around column and table names.

[the remark this answers is not in the document]
Comment (orcunc) [reply extensions/mapstore/src/main/java/com/hazelcast/mapstore/Queries.java:46]: Ok. I have used the SqlDialect class to quote the identifiers

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/H2UpsertQueryBuilder.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+        stringBuilder.append("MERGE INTO ")
+                .append(jdbcTable.getExternalName())
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/H2UpsertQueryBuilder.java]: I have used the SqlDialect class to quote the identifiers

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/H2UpsertQueryBuilderTest.java (hunk 9 above)
Comment (orcunc) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/H2UpsertQueryBuilderTest.java]: Removed the whitespace after KEY. It will be
`KEY ("pk1","pk2")`

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/H2UpsertQueryBuilderTest.java (hunk 8 above)
Comment (orcunc) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/H2UpsertQueryBuilderTest.java:98]: Ok. I have deleted the empty lines in the methods

File: extensions/mapstore/src/test/java/com/hazelcast/mapstore/GenericMapStoreTest.java [hunk 11]
```diff
@@ -388,16 +388,16 @@ public void givenNoRows_whenLoadAllKeys_thenEmptyIterable() throws Exception {
 
     @Test
     public void whenStore_thenTableContainsRow() throws Exception {
-        createTable(mapName, "\"person-id\" INT PRIMARY KEY", "name VARCHAR(100)");
```
[the remark this answers is not in the document]
Comment (orcunc) [reply extensions/mapstore/src/test/java/com/hazelcast/mapstore/GenericMapStoreTest.java:391]: Ok I have reverted everything back to person-id

File: extensions/mapstore/src/test/java/com/hazelcast/mapstore/GenericMapStoreMSSQLUpsertTest.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        assertJdbcRowsAnyOrder(tableName,
+                new Row(0, "name-0")
+        );
+
+        map.put(0, new Person(0, "updated"));
+
+        assertJdbcRowsAnyOrder(tableName,
+                new Row(0, "updated")
+        );
+    }
```
[the remark this answers is not in the document]
Comment (orcunc) [reply extensions/mapstore/src/test/java/com/hazelcast/mapstore/GenericMapStoreMSSQLUpsertTest.java:105]: Ok. I have subclassed the tests

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/MappingJdbcSqlConnectorTest.java [hunk 13]
```diff
@@ -94,12 +93,8 @@ public void createMappingTableDoesNotExist() {
                         + " '" + OPTION_EXTERNAL_DATASTORE_REF + "'='" + TEST_DATABASE_REF + "'"
                         + ")"
         ))
-                .isInstanceOf(HazelcastSqlException.class)
-                .is(Assertions.anyOf(
-                        hasMessage("Table \"" + tableName + "\" not found"), // H2
-                        hasMessage("relation \"" + tableName + "\" does not exist"), // Postgres
-                        hasMessage("Table 'test." + tableName + "' doesn't exist") // MySQL
-                ));
```
[the remark this answers is not in the document]
Comment (orcunc) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/MappingJdbcSqlConnectorTest.java:102]: JdbcSqlConnector.readDbFields() is reading the table fields with JDBC. There is no DB specific error message now

File: extensions/mapstore/src/test/java/com/hazelcast/mapstore/GenericMapStoreIntegrationTest.java [hunk 14]
```diff
@@ -215,32 +213,7 @@ public void testDestroy() {
 
         map.destroy();
 
-        assertTrueEventually(() -> {
-            assertRowsAnyOrder(client, "SHOW MAPPINGS", newArrayList());
-        }, 5);
-    }
-
-    @Test
-    public void testPutWithColumnMismatch() {
```
[the remark this answers is not in the document]
Comment (orcunc) [reply extensions/mapstore/src/test/java/com/hazelcast/mapstore/GenericMapStoreIntegrationTest.java:224]: Because the primary key is coming from the method call not from GenericRecord

File: extensions/mapstore/pom.xml [hunk 15]
```diff
@@ -101,5 +101,31 @@
...
+            <version>${testcontainers.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.microsoft.sqlserver</groupId>
+            <artifactId>mssql-jdbc</artifactId>
+            <version>11.2.2.jre8</version>
+            <scope>test</scope>
+        </dependency>
+        <!--MariDB-->
```
[the remark this answers is not in the document]
Comment (orcunc) [reply extensions/mapstore/pom.xml:117]: Ok

File: extensions/mapstore/src/test/java/com/hazelcast/mapstore/GenericMapStoreMSSQLUpsertTest.java [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+        putMap.put(45, new Person(45, "45"));
+        map.putAll(putMap);
+
+        assertJdbcRowsAnyOrder(tableName,
+                new Row(0, "0"),
+                new Row(42, "42"),
+                new Row(44, "44"),
+                new Row(45, "45")
+        );
+    }
```
[the remark this answers is not in the document]
Comment (orcunc) [reply extensions/mapstore/src/test/java/com/hazelcast/mapstore/GenericMapStoreMSSQLUpsertTest.java]: Ok. I have subclassed the tests

**Comment by orcunc** [comment]
As @frant-hartm has requested, I have split the PR into two. This one is only about generating the SQLs. GenericMapStore related changes are reverted and they will be in another PR

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java [hunk 17]
```diff
@@ -142,34 +145,52 @@ private Map<String, DbField> readDbFields(
...
-            for (int i = 1; i <= metaData.getColumnCount(); i++) {
-                String columnName = metaData.getColumnName(i);
+        } catch (Exception e) {
+            throw new HazelcastException("Could not execute readDbFields for table " + externalTableName, e);
+        } finally {
+            closeDataSource(dataSource);
+        }
+    }
+
+    private static Map<String, DbField> readColumns(String externalTableName, DatabaseMetaData databaseMetaData,
```
Comment (frant-hartm) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java:163]: The `readPrimaryKeyColumns` is called first, so should be reordered with `readColumns`

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java [hunk 18]
```diff
@@ -372,6 +380,43 @@ public Vertex deleteProcessor(@Nonnull DAG dag, @Nonnull Table table0) {
...
+    // Returns if upsert is supported for the given dialect
+    protected boolean isUpsertSupported(JdbcTable jdbcTable) {
+        boolean result = false;
+        SqlDialect dialect = jdbcTable.sqlDialect();
+        if (dialect instanceof MysqlSqlDialect ||
+            dialect instanceof PostgresqlSqlDialect ||
+            dialect instanceof H2SqlDialect) {
+            result = true;
+        }
+        return result;
```
Comment (frant-hartm) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java:392]:
```suggestion
        SqlDialect dialect = jdbcTable.sqlDialect();
        return dialect instanceof MysqlSqlDialect ||
            dialect instanceof PostgresqlSqlDialect ||
            dialect instanceof H2SqlDialect;
```
?

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpsertBuilder.java [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+        return supportedDialects.stream().anyMatch(clazz -> clazz.isInstance(dialect));
+
+    }
+
+    static String getUpsertStatement(JdbcTable jdbcTable) {
+        SqlDialect sqlDialect = jdbcTable.sqlDialect();
+
+        String query = null;
+        if (sqlDialect instanceof MysqlSqlDialect) {
+            MySQLUpsertQueryBuilder builder = new MySQLUpsertQueryBuilder(jdbcTable, sqlDialect);
```
Comment (frant-hartm) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpsertBuilder.java:50]: We could pass only the jdbcTable, it has the dialect inside.

**Review by frant-hartm — CHANGES_REQUESTED** [review]
We are missing integration test for the `SINK INTO ...` see e.g. com.hazelcast.jet.sql.impl.connector.jdbc.InsertJdbcSqlConnectorTest and the sub-classes.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java (hunk 18 above)
Comment (orcunc) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java]: Ok and moved the code to UpsertBuilder. Since it is more relevant in that class

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java (hunk 17 above)
Comment (orcunc) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/JdbcSqlConnector.java:173]: Ok

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpsertBuilder.java (hunk 19 above)
Comment (orcunc) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpsertBuilder.java]: Ok

**Comment by orcunc** [comment]
Added test to InsertJdbcSqlConnectorTest for SINK INTO

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/SinkIntoJdbcSqlConnectorTest.java [hunk 20]
```diff
@@ -0,0 +1,169 @@
...
+        createTable(tableName);
+        createMapping(tableName);
+
+        execute("SINK INTO " + tableName + " VALUES (0, 'name-0')");
+
+        assertJdbcRowsAnyOrder(tableName, new Row(0, "name-0"));
+    }
+
+    @Test
+    public void insertIntoTableWithExternalName() throws Exception {
```
Comment (frant-hartm) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/SinkIntoJdbcSqlConnectorTest.java:51]: Test method name is wrong. Fix all methods.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/SinkIntoJdbcSqlConnectorTest.java [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+
+        execute("SINK INTO " + tableName + " (name, id) VALUES ('name-0', 0)");
+
+        assertJdbcRowsAnyOrder(tableName,
+                new Row(0, "name-0")
+        );
+    }
+
+    @Test
+    public void sinkIntoTableWithColumns() throws Exception {
```
Comment (frant-hartm) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/SinkIntoJdbcSqlConnectorTest.java:151]: The name should reflect that this tests an update.

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/SinkIntoJdbcSqlConnectorTest.java [hunk 22]
```diff
@@ -0,0 +1,169 @@
...
+        createTable(tableName);
+        createMapping(tableName);
+
+        execute("INSERT INTO " + tableName + " (name, id) VALUES ('name-0', 0), ('name-1', 1)");
```
Comment (frant-hartm) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/SinkIntoJdbcSqlConnectorTest.java:155]: Ideally, we want to test just a small part of the functionality. Here a failure in INSERT will break this test, while the sink might work.

See how it's done e.g. in the test for UPDATE on line 42

https://github.com/hazelcast/hazelcast/blob/26fee11245a7d88ae5aada0ee84f30505e34e7e4/hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/UpdateJdbcSqlConnectorTest.java#L41-L61

**Review by frant-hartm — CHANGES_REQUESTED** [review]

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/SinkIntoJdbcSqlConnectorTest.java (hunk 20 above)
Comment (orcunc) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/SinkIntoJdbcSqlConnectorTest.java:51]: Ok

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/SinkIntoJdbcSqlConnectorTest.java (hunk 22 above)
Comment (orcunc) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/SinkIntoJdbcSqlConnectorTest.java]: Ok

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/SinkIntoJdbcSqlConnectorTest.java (hunk 21 above)
Comment (orcunc) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/connector/jdbc/SinkIntoJdbcSqlConnectorTest.java]: Renamed as updateTableWithColumns

**Comment by frant-hartm** [comment]
run-lab-run

**Review by frant-hartm — APPROVED** [review]

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/MySQLUpsertQueryBuilder.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (viliam-durina) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/jdbc/MySQLUpsertQueryBuilder.java]: sorry, my bad :)

**Comment by viliam-durina** [comment]
Now since we started quoting the external name in the generated queries, we can no longer access tables in non-default schema, because the `'.'` will also be quoted and interpreted literally. I think we have to rethink the external name here. I guess the external name should be a valid optionally-quoted SQL identifier. It should be valid according to the target database. Otherwise we can't have a literal and non-literal dot to delineate the schema. Or some other mechanism.

**Review by viliam-durina — APPROVED** [review]
Approving since the outstanding issue is moved to a follow-up issue here: https://hazelcast.atlassian.net/browse/HZ-2035
